server/storage/endpoint/gc_key_space.go

```diff
@@ -0,0 +1,174 @@
+	"math"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"go.etcd.io/etcd/clientv3"
+)
+
+// KeySpaceGCSafePoint is gcWorker's safepoint for specific key-space
+type KeySpaceGCSafePoint struct {
+	SpaceID   string `json:"space_id"`
+	SafePoint uint64 `json:"safe_point"`
+}
+
+// KeySpaceGCSafePointStorage defines the storage operations on KeySpaces' safe points
+type KeySpaceGCSafePointStorage interface {
+	// Service safe point interfaces.
+	SaveServiceSafePointByKeySpace(spaceID string, ssp *ServiceSafePoint) error
+	LoadServiceSafePointByKeySpace(spaceID, serviceID string) (*ServiceSafePoint, error)
+	LoadMinServiceSafePointByKeySpace(spaceID string, now time.Time) (*ServiceSafePoint, error)
+	RemoveServiceSafePointByKeySpace(spaceID, serviceID string) error
+	// GC safe point interfaces.
+	SaveGCSafePointByKeySpace(gcSafePoint *KeySpaceGCSafePoint) error
+	LoadGCSafePointByKeySpace(spaceID string) (*KeySpaceGCSafePoint, error)
+	LoadAllKeySpaceGCSafePoints(withGCSafePoint bool) ([]*KeySpaceGCSafePoint, error)
+}
+
+var _ KeySpaceGCSafePointStorage = (*StorageEndpoint)(nil)
+
+// SaveServiceSafePointByKeySpace saves service safe point under given key-space.
+func (se *StorageEndpoint) SaveServiceSafePointByKeySpace(spaceID string, ssp *ServiceSafePoint) error {
+	if ssp.ServiceID == "" {
+		return errors.New("service id of service safepoint cannot be empty")
+	}
+	key := KeySpaceServiceSafePointPath(spaceID, ssp.ServiceID)
+	value, err := json.Marshal(ssp)
+	if err != nil {
+		return err
+	}
+	return se.Save(key, string(value))
+}
+
+// LoadServiceSafePointByKeySpace reads ServiceSafePoint for the given key-space ID and service name.
+// Return nil if no safepoint not exist.
+func (se *StorageEndpoint) LoadServiceSafePointByKeySpace(spaceID, serviceID string) (*ServiceSafePoint, error) {
+	value, err := se.Load(KeySpaceServiceSafePointPath(spaceID, serviceID))
+	if err != nil || value == "" {
+		return nil, err
+	}
+	ssp := &ServiceSafePoint{}
+	if err := json.Unmarshal([]byte(value), ssp); err != nil {
+		return nil, err
+	}
+	return ssp, nil
+}
+
+// LoadMinServiceSafePointByKeySpace returns the minimum safepoint for the given key-space.
+// Note that gc worker safe point are store separately.
+// If no service safe point exist for the given key-space or all the service safe points just expired, return nil.
+func (se *StorageEndpoint) LoadMinServiceSafePointByKeySpace(spaceID string, now time.Time) (*ServiceSafePoint, error) {
+	prefix := KeySpaceServiceSafePointPrefix(spaceID)
+	prefixEnd := clientv3.GetPrefixRangeEnd(prefix)
+	keys, values, err := se.LoadRange(prefix, prefixEnd, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	min := &ServiceSafePoint{SafePoint: math.MaxUint64}
+	for i, key := range keys {
+		ssp := &ServiceSafePoint{}
+		if err = json.Unmarshal([]byte(values[i]), ssp); err != nil {
+			return nil, err
+		}
+
+		// remove expired safe points.
+		if ssp.ExpiredAt < now.Unix() {
+			err = se.Remove(key)
```

Comment:
Could you please Remove the expired items in a goroutine? Remove here in the main logic seems too expensive